File: domain/src/main/kotlin/com/aiassistant/domain/model/Persona.kt
/*
 * ============================================================
 * Android AI Assistant (Enterprise Edition)
 * ============================================================
 * Module     : domain
 * File       : Persona.kt
 * Purpose    : Persona — domain module component
 *
 * Architecture Layer : Domain
 * Pattern Used       : Kotlin Data Class
 *
 * Key Concepts:
 *   - Clean Architecture with strict layer separation
 *   - Hilt dependency injection
 *
 * Dependencies:
 *   - See import statements below
 * ============================================================
 */

/**
 * Persona.kt
 *
 * Purpose: Domain entity representing an AI persona with customizable system prompt and tone.
 * Architecture: domain module — pure Kotlin, zero Android or third-party framework dependencies.
 * Dependencies: None
 *
 * Requirements: 32.1, 32.3, 32.4, 32.5, 32.6
 */

package com.aiassistant.domain.model

/**
 * The tone/style that the AI should adopt when using this persona.
 *
 * Maps to the "tone" string field in the backend and Room entity.
 */
enum class PersonaTone(val value: String) {
    /** Professional, formal, and business-appropriate language. */
    PROFESSIONAL("professional"),

    /** Relaxed, conversational, and friendly language. */
    CASUAL("casual"),

    /** Brief, to-the-point responses with minimal elaboration. */
    CONCISE("concise"),

    /** Thorough, comprehensive responses with extensive explanation. */
    DETAILED("detailed"),

    /** Imaginative, exploratory, and creative language. */
    CREATIVE("creative");

    companion object {
        fun fromValue(value: String): PersonaTone = entries.firstOrNull { it.value == value } ?: PROFESSIONAL
    }
}

/**
 * Represents an AI persona with customizable system prompt and tone.
 *
 * THE AI_Orchestrator SHALL inject the selected Persona's system prompt + tone + scope
 * into the LLM system message (Requirement 32.8).
 *
 * @param id                Unique identifier for the persona.
 * @param userId            Identifier of the persona's owner.
 * @param name              Display name of the persona (1–80 characters).
 * @param systemPrompt      Custom system prompt text (1–4,000 characters).
 * @param tone              The tone/style for AI responses using this persona.
 * @param scopeDescription  Optional description of when to use this persona (0–500 characters).
 * @param adminLocked       When true, only admin users can edit or delete this persona.
 * @param allowedRoles      List of role identifiers permitted to use this persona (empty = all roles).
 * @param createdAt         Epoch milliseconds when the persona was created.
 * @param updatedAt         Epoch milliseconds of the last edit.
 */
data class Persona(
    val id: String,
    val userId: String,
    val name: String,
    val systemPrompt: String,
    val tone: PersonaTone = PersonaTone.PROFESSIONAL,
    val scopeDescription: String? = null,
    val adminLocked: Boolean = false,
    val allowedRoles: List<String> = emptyList(),
    val createdAt: Long,
    val updatedAt: Long
)

File: backend/app/models/persona.py
# ============================================================
# Android AI Assistant (Enterprise Edition) — Backend
# ============================================================
# Module  : models
# File    : persona.py
# Purpose : persona — models module
#
# Architecture Layer : ORM Model
# Pattern Used       : Python Module
#
# Key Concepts:
#   - FastAPI async request handling
#   - SQLAlchemy 2.x async ORM
#
# Dependencies:
#   - See import statements below
# ============================================================

"""ORM model for the ``personas`` table.

The ``Persona`` model stores custom AI personas that users can create to customize
the AI's personality, system prompt, tone, and scope. Each user can have up to 20
personas. Admin users can create admin-locked personas that are shared across
specific roles.

Security notes
--------------
- ``system_prompt`` is validated for prompt injection patterns before persistence
  (handled in the service layer via SafetyService).
- ``admin_locked`` personas cannot be edited or deleted by non-admin users.
- ``allowed_roles`` controls which RBAC roles can see and use the persona.

Requirements: 32.1, 32.2, 32.3, 32.4, 32.5, 32.6
"""

from __future__ import annotations

import enum
import uuid
from typing import TYPE_CHECKING

from sqlalchemy import Boolean, Enum, ForeignKey, String, Text
from sqlalchemy.dialects.postgresql import ARRAY
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import Base, TimestampMixin, uuid_pk

if TYPE_CHECKING:
    from app.models.user import User


class PersonaTone(str, enum.Enum):
    """Tone options for persona behavior."""

    professional = "professional"
    casual = "casual"
    concise = "concise"
    detailed = "detailed"
    creative = "creative"


class Persona(Base, TimestampMixin):
    """SQLAlchemy ORM model representing a custom AI persona.

    Relationships
    -------------
    - Many-to-one with ``User`` (creator/owner).
    """

    __tablename__ = "personas"

    id: Mapped[uuid.UUID] = uuid_pk()
    user_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
        comment="User who created this persona; admin-locked personas still have an owner",
    )
    name: Mapped[str] = mapped_column(
        String(80),
        nullable=False,
        comment="Persona display name (1-80 characters)",
    )
    system_prompt: Mapped[str] = mapped_column(
        Text,
        nullable=False,
        comment="Custom system prompt injected into AI context (1-4,000 characters)",
    )
    tone: Mapped[PersonaTone] = mapped_column(
        Enum(PersonaTone, name="persona_tone", create_type=True),
        nullable=False,
        default=PersonaTone.professional,
        comment="Tone/style of the AI response",
    )
    scope_description: Mapped[str] = mapped_column(
        String(500),
        nullable=False,
        default="",
        server_default="",
        comment="Optional description of when to use this persona (0-500 characters)",
    )
    admin_locked: Mapped[bool] = mapped_column(
        Boolean,
        nullable=False,
        default=False,
        server_default="false",
        comment="When True, only admins can edit/delete; visible to allowed_roles",
    )
    allowed_roles: Mapped[list[str]] = mapped_column(
        ARRAY(String),
        nullable=False,
        default=list,
        server_default="{}",
        comment="RBAC roles allowed to view/use this persona; empty list = owner-only",
    )

    # ------------------------------------------------------------------
    # Relationships (back-populated in parent models)
    # ------------------------------------------------------------------
    user: Mapped[User] = relationship("User", back_populates="personas")

    def __repr__(self) -> str:
        return (
            f"<Persona id={self.id!s} name={self.name!r} "
            f"tone={self.tone.value!r} admin_locked={self.admin_locked}>"
        )
